geometric_kernels/spaces/so.py

```diff
@@ -0,0 +1,368 @@
+
+
+def partition_dominance_cone(partition):
+    '''
+    Calculates partitions dominated by a given one
+    and having the same number of parts (including zero parts of the original)
+    '''
+    cone = {partition}
+    new_partitions = {0}
+    prev_partitions = cone
+    while new_partitions:
+        new_partitions = set()
+        for partition in prev_partitions:
+            for i in range(len(partition) - 1):
+                if partition[i] > partition[i + 1]:
+                    for j in range(i + 1, len(partition)):
+                        if partition[i] > partition[j] + 1 and partition[j] < partition[j - 1]:
+                            new_partition = list(partition)
+                            new_partition[i] -= 1
+                            new_partition[j] += 1
+                            new_partition = tuple(new_partition)
+                            if new_partition not in cone:
+                                new_partitions.add(new_partition)
+        cone.update(new_partitions)
+        prev_partitions = new_partitions
+    return cone
+
+
+def partition_dominance_or_subpartition_cone(partition):
+    '''
+        Calculates subpartitions and partitions dominated by a given one
+        and having the same number of parts (including zero parts of the original)
+        '''
+    cone = {partition}
+    new_partitions = {0}
+    prev_partitions = cone
+    while new_partitions:
+        new_partitions = set()
+        for partition in prev_partitions:
+            for i in range(len(partition) - 1):
+                if partition[i] > partition[i + 1]:
+                    new_partition = list(partition)
+                    new_partition[i] -= 1
+                    new_partition = tuple(new_partition)
+                    if new_partition not in cone:
+                        new_partitions.add(new_partition)
+                    for j in range(i + 1, len(partition)):
+                        if partition[i] > partition[j] + 1 and partition[j] < partition[j - 1]:
+                            new_partition = list(partition)
+                            new_partition[i] -= 1
+                            new_partition[j] += 1
+                            new_partition = tuple(new_partition)
+                            if new_partition not in cone:
+                                new_partitions.add(new_partition)
+        cone.update(new_partitions)
+        prev_partitions = new_partitions
+    return cone
+
+
+class SOEigenfunction(LieGroupAddtitionTheorem):
+    def __init__(self, num_levels, n, init_eigenfunctions=True):
+        self.n = n
+        self.dim = n * (n-1) // 2
+        self.rank = n // 2
+
+        if self.n % 2 == 0:
+            self.rho = np.arange(self.rank - 1, -1, -1)
+        else:
+            self.rho = np.arange(self.rank - 1, -1, -1) + 0.5
+
+        self._num_levels = num_levels
+        self._signatures = self.generate_signatures(self._num_levels) # both have the length num_levels
```

Comment:
Again, there seems a typo with the underscore? Also, now there is only one return value, while the comment still refers to "both" of something. Please clarify

Reply:
There is no type with an underscore, below getter function `num_levels()` returns this value.